File: LV1/1.8/nineknights/nineknights.rb
def vaild_bound?(i, j)
  i >= 0 && j >= 0 && i < 5 && j < 5
end

def gen_new_pos(i, j)
  pos = [[-1, 2], [1, 2], [-1, -2], [1, 2], [2, -1], [2, 1], [-2, -1], [-2, 1]]
  pos.map { |ni, nj| [i + ni, j + nj] }
end

def check_pos(board, i, j)
  gen_new_pos(i, j).each do |ni, nj|
    next unless vaild_bound?(ni, nj)
    return true if board[ni][nj] == 'k'
  end
  false
end

def can_attack(board)
  [*0..4].product([*0..4]).each do |i, j|
    if board[i][j] == 'k'
      return true if check_pos(board, i, j)
    end
  end
  false
end

board = (1..5).map { gets }
n_k = board.map { |l| l.count('k') }.sum

is_valid = n_k == 9
is_valid = !can_attack(board) if is_valid

puts is_valid ? 'valid' : 'invalid'

File: LV1/1.8/nineknights/nineknights.py
def is_vaild_bound(i, j):
    return i >= 0 and j >= 0 and i < 5 and j < 5


def check_pos(board, i, j):
    for d in pos:
        ni = i + d[0]
        nj = j + d[1]

        if not is_vaild_bound(ni, nj):
            continue
        if board[ni][nj] == 'k':
            return True
    return False


def can_attack(board):
    for i in range(5):
        for j in range(5):
            if board[i][j] == 'k':
                if check_pos(board, i, j):
                    return True
    return False


pos = [(-1, 2), (1, 2), (-1, -2), (1, 2),
       (2, -1), (2, 1), (-2, -1), (-2, 1)]

board = [input() for _ in range(5)]
n_k = sum([l.count('k') for l in board])

is_valid = n_k == 9

if is_valid:
    is_valid = not can_attack(board)

if is_valid:
    print("valid")
else:
    print("invalid")
